libsplinter/src/service/scabbard/state.rs

```diff
@@ -609,22 +609,44 @@ impl BatchHistory {
 
     fn update_batch_status(&mut self, signature: &str, status: BatchStatus) {
         match self.history.get_mut(signature) {
-            Some(ref mut batch) if batch.status == BatchStatus::Pending => {
-                batch.set_status(status);
+            Some(info) => {
+                if info.status == BatchStatus::Pending {
+                    info.set_status(status);
+                } else {
+                    error!(
+                        "Received status update for batch that was not pending: {:?}",
+                        signature
+                    );
+                }
             }
-            _ => (),
-        };
+            None => {
+                debug!(
+                    "Received status update for batch that is not in the history: {:?}",
+                    signature
+                );
+            }
+        }
     }
 
     fn commit(&mut self, signature: &str) {
-        let info = if let Some(info) = self.history.get_mut(signature) {
-            info
-        } else {
-            return;
-        };
-
-        if let BatchStatus::Valid(t) = info.status.clone() {
-            info.set_status(BatchStatus::Committed(t));
+        match self.history.get_mut(signature) {
+            Some(info) => match info.status.clone() {
```

Comment:
This clone could be done as late as possible - i.e. when you're going to return set the status

Reply:
That actually produces a warning:

warning: cannot borrow `*info` as mutable because it is also borrowed as immutable
   --> libsplinter/src/service/scabbard/state.rs:634:21
    |
632 |             Some(info) => match &info.status {
    |                                 ------------ immutable borrow occurs here
633 |                 BatchStatus::Valid(txns) => {
634 |                     info.set_status(BatchStatus::Committed(txns.clone()));
    |                     ^^^^ mutable borrow occurs here        ---- immutable borrow later used here
    |
    = note: `#[warn(mutable_borrow_reservation_conflict)]` on by default
    = warning: this borrowing pattern was not meant to be accepted, and may become a hard error in the future
    = note: for more information, see issue #59159 <https://github.com/rust-lang/rust/issues/59159>